templates/static/js/example.js

```diff
@@ -79,136 +140,232 @@
+   * @param {HTMLElement} placementElement The placeholder element for the example
+   * @returns {Promise<String>} The Jinja content block of the example
+   */
+  async function fetchJinjaContentBlock(placementElement) {
+    // Raw templates are not served at standalone paths, so strip it from the URL if it was found.
+    const exampleUrl = new URL(`${placementElement.href.replace(/standalone/, '/')}`);
 
+    // Add `?raw=true` query parameter to the URL to request the raw Jinja template
+    const queryParams = new URLSearchParams(exampleUrl.search);
+    queryParams.set('raw', true);
+    exampleUrl.search = queryParams.toString();
+
+    const rawJinjaTemplate = await fetchResponseText(exampleUrl.toString());
+    return formatSource(getExampleSection('jinja', rawJinjaTemplate), 'jinja');
+  }
+
+  /**
+   * Replaces an example placeholder element with its rendered result and code snippet.
+   * @param {HTMLAnchorElement} placementElement `a.js-example` element used as a placeholder for the example to render
+   * @param {String} renderedHtml Full document HTML of the example as it is shown to end-users
+   * @param {String|null} jinjaTemplate Jinja Template of the example as it may be used by developers, if supported
+   */
+  async function renderExample(placementElement) {
+    const codeSnippet = document.createElement('div');
     codeSnippet.classList.add('p-code-snippet', 'is-bordered');
 
-    var header = document.createElement('div');
+    const header = document.createElement('div');
     header.classList.add('p-code-snippet__header');
-    var titleEl = document.createElement('h5');
+
+    const titleEl = document.createElement('h5');
     titleEl.classList.add('p-code-snippet__title');
 
-    // example page title is structured as "... | Examples | Vanilla documentation"
-    // we want to strip anything after first | pipe
-    titleEl.innerText = title.split('|')[0];
+    const srcData = {
+      html: undefined,
+      jinja: undefined,
+      css: undefined,
+      js: undefined,
+      codePen: undefined,
+    };
 
-    header.appendChild(titleEl);
-    codeSnippet.appendChild(header);
+    const exampleRequests = [];
+
+    const fetchHtml = fetchHtmlSource(placementElement).then(({rendered: renderedHtml, body: bodyHtml}) => {
+      const title = getExampleSection('title', renderedHtml).split('|')[0];
+      const headHtml = getExampleSection('head', renderedHtml);
+      const htmlBodySource = formatSource(stripScriptsFromSource(bodyHtml), 'html');
+      const jsSource = formatSource(getScriptFromSource(bodyHtml), 'js');
+      const cssSource = formatSource(getStyleFromSource(headHtml), 'css');
+      const externalScripts = getExternalScriptsFromSource(renderedHtml);
+
+      titleEl.innerText = title;
+      header.appendChild(titleEl);
+      codeSnippet.appendChild(header);
+      placementElement.parentNode.insertBefore(codeSnippet, placementElement);
+
+      // HTML iframe is required, so throw if it fails
```

Comment:
It is a bit confusing how now we render the iframe right away, but not put anything in it, just wait for the promises to be all resolved first.

I kind of understand that this comes from building on top of previous (simpler single-fetch approach), and it's not immediately apparent to me how this could work differently.

Overall it all works fine as far as I understand, but it's a bit hard to reason what happens when.

Reply:
@bartaz 

I've revised this somewhat to move all of the rendering logic outside of the async part of this operation - now there is a `srcData` object that the two promises asynchronously populate, then rendering occurs after both promises are resolved.

This does mean that it's possible for the iframe's HTML to be ready and available, but not rendered due to waiting for resolution of the raw endpoint for the Jinja template. 

The most problematic function that I'm trying to avoid calling in async blocks is `Node:appendChild()`, which is very order-sensitive. There can only be `appendChild()` in one of the two promises, as otherwise they'd be racing each other. Currently I have no `appendChild()` calls in either of the promises to make the code read nicer, but we could move the `iframe` rendering into the HTML promise so that it doesn't need to wait for the Jinja promise to resolve. 

What do you think?